src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -1126,7 +1147,49 @@ bool AdaptDispatch::SetCursorKeysMode(const bool applicationMode)
 // - True if handled successfully. False otherwise.
 bool AdaptDispatch::EnableCursorBlinking(const bool enable)
 {
-    return _pConApi->EnableCursorBlinking(enable);
+    auto& cursor = _pConApi->GetTextBuffer().GetCursor();
+    cursor.SetBlinkingAllowed(enable);
+
+    // GH#2642 - From what we've gathered from other terminals, when blinking is
+    // disabled, the cursor should remain On always, and have the visibility
+    // controlled by the IsVisible property. So when you do a printf "\e[?12l"
+    // to disable blinking, the cursor stays stuck On. At this point, only the
+    // cursor visibility property controls whether the user can see it or not.
+    // (Yes, the cursor can be On and NOT Visible)
+    cursor.SetIsOn(true);
+
+    return !_pConApi->IsConsolePty();
+}
+
+// Routine Description:
+// - Internal logic for adding or removing lines in the active screen buffer.
+//   This also moves the cursor to the left margin, which is expected behavior for IL and DL.
+// Parameters:
+// - delta - Number of lines to modify (positive if inserting, negative if deleting).
+// Return Value:
+// - <none>
+void AdaptDispatch::_InsertDeleteLineHelper(const int32_t delta)
+{
+    const auto viewport = _pConApi->GetViewport();
+    auto& textBuffer = _pConApi->GetTextBuffer();
+    const auto bufferWidth = textBuffer.GetSize().Width();
+
+    auto& cursor = textBuffer.GetCursor();
+    const auto row = cursor.GetPosition().Y;
+
+    const auto [topMargin, bottomMargin] = _GetVerticalMargins(viewport, true);
+    if (row >= topMargin && row <= bottomMargin)
+    {
+        // We emulate inserting and deleting by scrolling the area between the cursor and the bottom margin.
+        _ScrollRectVertically(textBuffer, { 0, row, bufferWidth, bottomMargin + 1 }, delta);
+
+        // The IL and DL controls are also expected to move the cursor to the left margin.
+        // For now this is just column 0, since we don't yet support DECSLRM.
+        cursor.SetXPosition(0);
+        cursor.SetDelay(false);
+        cursor.SetIsOn(true);
+        cursor.SetHasMoved(true);
```

Comment:
Yeah. I was considering that, but I thought maybe we could push that off for a follow up issue because I think this cursor handling needs further investigation.

If these calls are needed all the time, then why aren't they set automatically in the `SetPosition` method? If they aren't needed all the time, then we may need to be more selective about where they are used, and dumping these calls in a helper is just hiding the problem.

And note that Windows Terminal doesn't use these flags at all as far as I can tell, so I'm hoping it might be possible to do away with them altogether.

Reply:
FYI, I eventually decided you were right about this and added a helper for setting these flags. It does make things a bit tidier and gave me the opportunity to add a bit of documentation on what we are doing here (at least as I understand it). See commit 446002ed47d2e1a85f256362689435682caa5558.